### PR TITLE
Reduce copy overhead when writing metric events

### PR DESCRIPTION
**Pull request by phijor**
Previously, `otf2::event::metric` stored its metric values as pairs of
referenced metric members and metric values.  This incurred overhead
when writing the event to a trace because of the following reasons:

1.  `OTF2_EvtWriter_Metric` expects two continuous arrays, one for
        metric values and one for the corresponding `OTF2_TypeID`s.
        Each of them had to be created and copied from the pairs
        contained in a metric event.

2.  Each type ID had to be looked up in the referenced metric
        member.

With this commit, `otf2::event::metric` stores both metric values and type
IDs directly in two continuous buffers which.  The contents of these
buffers is copied exactly once from the OTF2 interface; when writing
an event to a trace, no additional copy operations are required.

It is still possible to obtain a `otf2::event::metric::value_container`
from a metric; they are created on-demand an will still contain a
refercence to metric member.

NOTE: this commit makes **BREAKING CHANGES** as the constructor for metric
events has changed.

### CODE REVIEW

File: include/otf2xx/event/metric.hpp [hunk 1]
```diff
@@ -165,34 +172,68 @@ namespace event
             }
         };
 
+        class metric_values
+        {
+        public:
+            template <typename Definition>
+            using weak_ref = otf2::definition::detail::weak_ref<Definition>;
+
+            metric_values(const OTF2_Type* type_ids, const OTF2_MetricValue* values,
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: I don't like that constructor at all. I'd suggest passing std::vector as &&. If you really insist, take them as forwarding (aka. universal) reference

File: include/otf2xx/event/metric.hpp [hunk 2]
```diff
@@ -165,34 +172,68 @@ namespace event
...
+                          std::size_t num_events)
+            : type_ids_(type_ids, type_ids + num_events), values_(values, values + num_events)
+            {
+            }
+
+            std::size_t size() const
+            {
+                assert(type_ids_.size() == values_.size());
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: This check doesn't make much sense here. Move it into the constructor.

File: include/otf2xx/event/metric.hpp [hunk 3]
```diff
@@ -203,44 +244,47 @@ namespace event
...
         }
 
-        const std::vector<value_container>& values() const
+        const metric_values& values() const
         {
             return values_;
         }
 
-        const value_container& get_value_for(const otf2::definition::metric_member& member) const
+        value_container get_value_for(const otf2::definition::metric_member& member) const
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: When I'm looking at this function now, I think, we should change the semantics of the` value_container` a bit. It should work like a proxy object. That means, the `value_container` should store a reference to the OTF2_MetricValue, OTF2_Type and otf2::definition::metric_class. And it should be able to set the value and the type. In case of type, update the reference OTF2_Type *and* the metric_member::type(). When you're on it, rename it to `value_proxy`.

File: include/otf2xx/event/metric.hpp [hunk 4]
```diff
@@ -267,12 +311,28 @@ namespace event
             metric_instance_ = mi;
         }
 
+        weak_ref<otf2::definition::metric_class> resolve_metric_class() const
+        {
+            assert(((void)"Malformed metric event: does reference neither metric class nor metric "
+                          "instance!",
+                    has_metric_class() || has_metric_instance()));
+
+            if (has_metric_class())
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: Do the if the other way around; guard the dereference of metric_instance.

Reason: without the assertion, it would still not cause a segfault and the calling context could potentially cope with a weak_ref to a nullptr.

File: include/otf2xx/event/metric.hpp [hunk 5]
```diff
@@ -267,12 +311,28 @@ namespace event
             metric_instance_ = mi;
         }
 
+        weak_ref<otf2::definition::metric_class> resolve_metric_class() const
+        {
+            assert(((void)"Malformed metric event: does reference neither metric class nor metric "
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: No, we allow this. Sort of.  Look below.

File: include/otf2xx/writer/local.hpp [hunk 6]
```diff
@@ -174,19 +174,12 @@ namespace writer
...
-
-            type_ids_.resize(num_members);
-            values_.resize(num_members);
-
-            for (std::size_t i = 0; i < num_members; i++)
-            {
-                type_ids_[i] = static_cast<OTF2_Type>(data.values()[i].metric->value_type());
-                values_[i] = data.values()[i].value;
-            }
+            const auto& type_ids = data.values().type_ids();
```
Comment (bmario) [inline include/otf2xx/writer/local.hpp]: These two local variables have no point. Just call it directly.

File: src/reader/callback/events.cpp [hunk 7]
```diff
@@ -142,48 +141,35 @@ namespace reader
                     return static_cast<OTF2_CallbackCode>(OTF2_SUCCESS);
                 }
 
-                typedef otf2::event::metric::value_container value_container;
+                otf2::chrono::time_point timestamp =
+                    otf2::chrono::convert(reader->ticks_per_second())(otf2::chrono::ticks(
```
Comment (bmario) [inline src/reader/callback/events.cpp:143]: In hindsight, this conversion code is disgusting. We need to make a separate PR and move all occurrences into a function.

File: src/reader/callback/events.cpp [hunk 8]
```diff
@@ -142,48 +141,35 @@ namespace reader
...
+                        time - reader->clock_properties().start_time().count()));
 
-                std::vector<value_container> values(numberOfMetrics);
+                otf2::event::metric::metric_values metric_values{ typeIDs, metricValues,
+                                                                  numberOfMetrics };
 
                 // assumes a valid trace file
                 if (reader->metric_classes().count(metric))
```
Comment (bmario) [inline src/reader/callback/events.cpp]: Ideally, this should be a call to find()

File: include/otf2xx/event/metric.hpp [hunk 9]
```diff
@@ -256,7 +300,7 @@ namespace event
             return static_cast<bool>(metric_instance_);
         }
 
-        otf2::definition::metric_instance metric_instance() const
+        weak_ref<otf2::definition::metric_instance> metric_instance() const
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: Add setters (`void metric_instance(weak_ref<otf2::definition::metric_instance>)`) for instance and class.

File: src/reader/callback/events.cpp [hunk 10]
```diff
@@ -142,48 +141,35 @@ namespace reader
...
                 }
 
-                typedef otf2::event::metric::value_container value_container;
+                otf2::chrono::time_point timestamp =
+                    otf2::chrono::convert(reader->ticks_per_second())(otf2::chrono::ticks(
+                        time - reader->clock_properties().start_time().count()));
 
-                std::vector<value_container> values(numberOfMetrics);
+                otf2::event::metric::metric_values metric_values{ typeIDs, metricValues,
+                                                                  numberOfMetrics };
```
Comment (bmario) [inline src/reader/callback/events.cpp]: Let's restructure this function:

Something around:
```
// this should be fine, as setting a weak_ref is only one pointer copy, so we can delay it
// however, it requires a new constructor
otf2::event::metric metric_event{ attributeList, timestamp, std::move(metric_values) };

if(auto it = reader->metric_classes().find(metric) != reader->metric_classes().end()) {
  metric_event.metric_class(*it);
}
else {
  metric_event.metric_instance(reader->metric_instances()[metric]);
}

reader->callback().event(reader->locations()[locationID], metric_event);
```

**Review by bmario — CHANGES_REQUESTED** [review]

File: include/otf2xx/writer/local.hpp (hunk 6 above)
Comment (phijor) [reply include/otf2xx/writer/local.hpp]: I would argue that while these local variables could be replaced by direct calls, it becomes a daunting task to parse the following call to `OTF2_EvtWriter_Metric`.

In my opinion, `data.values().values().data()` really just isn't a readable parameter, and passing `metric_values` instead helps understanding the semantics of this call.

File: include/otf2xx/event/metric.hpp [hunk 11]
```diff
@@ -203,36 +360,45 @@ namespace event
...
-        const value_container& get_value_for(const otf2::definition::metric_member& member) const
+        value_proxy get_value_for(const otf2::definition::metric_member& member)
         {
-            std::size_t i = 0;
+            const otf2::definition::metric_class& metric_class = this->resolve_metric_class();
 
-            auto mc = metric_class_.lock();
+            // TODO: maybe check if metric_class is undefined? This might happen
+            // if the event was constructed without a reference to a metric
+            // class or metric instance.
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: `assert(metric_class.is_valid());`

But keep the comment, it might prove useful.

File: include/otf2xx/event/metric.hpp [hunk 12]
```diff
@@ -203,36 +360,45 @@ namespace event
...
 
-            if (metric_instance_)
+            // Look up the index of a member inside of metric class and use it to
+            // construct a value_proxy from the right OTF2_Type and OTF2_MetricValue.
+            std::size_t index = 0;
+            for (const auto& class_member : metric_class)
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]:
```
auto it = std::find(metric_class.begin(), metric_class.end(), member);

if(it == metric_class.end())
{
    throw std::out_of_range("Failed to look up metric_member inside metric_class");
}

return std::distance(metric_class.begin(), it);
```

Flip a coin, if heads use this code.

File: include/otf2xx/event/metric.hpp [hunk 13]
```diff
@@ -136,63 +130,226 @@ namespace event
...
+        /**
+         * \brief A proxy class allowing type safe, correctly scaled access to a
+         *        metric value
+         */
+        class value_proxy
+        {
+            template <typename T>
+            T scale(T x) const
+            {
+                using value_base_type = otf2::definition::metric_member::value_base_type;
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: I like the split of the scaling and conversion. 🐝

**Review by bmario — CHANGES_REQUESTED** [review]
Just a few minor things. I'd like to merge this PR first, before working on #11

File: include/otf2xx/writer/local.hpp [hunk 14]
```diff
@@ -160,33 +160,26 @@ namespace writer
...
-            for (std::size_t i = 0; i < num_members; i++)
-            {
-                type_ids_[i] = static_cast<OTF2_Type>(data.values()[i].metric->value_type());
-                values_[i] = data.values()[i].value;
-            }
-
-            check(OTF2_EvtWriter_Metric(evt_wrt_, data.attribute_list().get(),
-                                        convert(data.timestamp()), metric, num_members,
-                                        type_ids_.data(), values_.data()),
+            check(OTF2_EvtWriter_Metric(evt_wrt_, metric.attribute_list().get(),
```
Comment (bmario) [inline include/otf2xx/writer/local.hpp:180]: When I think about this, I'm not convinced. The issue is that this code is indeed faster than the old one. However, it's not optimal in the only writing use case like lo2s.

The metric_values owns the memory which is owned by the event itself. Therefore you're forced to allocate new memory for each event. This is good enough for the rewrite like in HAEC-Sim and the trace merger, but not in lo2s. We need to talk about this in the next meeting.

Comment (phijor) [reply include/otf2xx/writer/local.hpp:180]: I've been testing the commits from this branch with `lo2s` and have already collected some ideas how remedy those issue. The basic idea is to pre-initialize a `otf2::event::metric` and update it when it needed, then writing it to the archive. This would remove a lot of the `TODO: avoid copy/shared pointers` comments while still using the new code.

Still, the interface of `otf2::event::metric::metric_values` needs some improvement. Especially construction with the correct type information from a `otf2::definition::metric_class` seems necessary.

Comment (bmario) [reply include/otf2xx/writer/local.hpp:180]: I've discussed this with @tilsche. We concluded that making the event mutable is sufficient.
So add a setter for time in event::base and expose the writable proxy from the metric event such that you can iterate it and directly change the value.

File: include/otf2xx/event/metric.hpp [hunk 15]
```diff
@@ -118,121 +120,428 @@ namespace event
...
         }
 
-        const value_container& get_value_for(const otf2::definition::metric_member& member) const
+        value_proxy get_value_at(std::size_t index)
         {
-            std::size_t i = 0;
+            auto metric_class = resolve_metric_class_ref();
 
-            auto mc = metric_class_.lock();
+            assert(metric_class->is_valid());
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: These cannot work, did you built in release?

The thing is, a weak_ref cannot check if it has a valid pointer. It can only be checked against nullptr

You'll probably want to check for `metric_class != nullptr` or something akin.

File: include/otf2xx/event/metric.hpp [hunk 16]
```diff
@@ -267,12 +576,33 @@ namespace event
...
+            }
+        }
+
+    private:
+        weak_ref<otf2::definition::metric_class> metric_class_;
+        weak_ref<otf2::definition::metric_instance> metric_instance_;
+        metric_values values_;
     };
 } // namespace event
 } // namespace otf2
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: Please split the file metric.hpp in 3 files:

- event::metric
- event::detail::type_value_proxy
- event::detail::metric_values

File: include/otf2xx/event/metric.hpp [hunk 17]
```diff
@@ -267,12 +576,33 @@ namespace event
...
         friend class otf2::writer::local;
 
     private:
-        otf2::definition::detail::weak_ref<otf2::definition::metric_class> metric_class_;
-        otf2::definition::detail::weak_ref<otf2::definition::metric_instance> metric_instance_;
-        std::vector<value_container> values_;
+        template <typename Definition>
+        using weak_ref = otf2::definition::detail::weak_ref<Definition>;
+
+        weak_ref<otf2::definition::metric_class> resolve_metric_class_ref() const
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: this shouldn't be called `resolve_metric_class_ref` to prevent confusions with `resolve_metric_class().ref()`

File: include/otf2xx/event/metric.hpp [hunk 18]
```diff
@@ -118,121 +120,428 @@ namespace event
...
+
+        value_proxy get_value_for(const otf2::definition::metric_member& member)
+        {
+            auto metric_class = this->resolve_metric_class_ref();
+
+            // TODO: maybe check if metric_class is undefined? This might happen
+            // if the event was constructed without a reference to a metric
+            // class or metric instance.
+
+            assert(metric_class->is_valid());
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: see comment above

File: include/otf2xx/event/metric.hpp [hunk 19]
```diff
@@ -118,121 +120,428 @@ namespace event
...
-            }
+            return value_proxy(values_[index], (*metric_class)[index]);
+        }
 
-            for (; i < mc.size(); i++)
+        const_value_proxy get_value_at(std::size_t index) const
+        {
+            auto metric_class = resolve_metric_class_ref();
+
+            assert(metric_class->is_valid());
```
Comment (bmario) [inline include/otf2xx/event/metric.hpp]: see comment above

**Review by bmario — CHANGES_REQUESTED** [review]

**Review by bmario — APPROVED** [review]
